### PR TITLE
fix: do not allow setting permissions for self

### PR DESCRIPTION
**Pull request by RiccardoM**
## Description
This PR changes how `MsgSetUserPermissions` is validated and how `MsgSetUserGroupPermissions` is handled to make sure a user cannot set their own permissions or the permissions of a group they are part of. 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://docs.cosmos.network/v0.44/building-modules/intro.html)
- [x] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/subspaces/keeper/msg_server.go [hunk 1]
```diff
@@ -276,6 +276,11 @@ func (k msgServer) SetUserGroupPermissions(goCtx context.Context, msg *types.Msg
 		return nil, sdkerrors.Wrapf(types.ErrPermissionDenied, "you cannot manage permissions in this subspace")
 	}
 
+	// Make sure that the user is not part of the group they want to change the permissions for
```
Comment (manu0466) [inline x/subspaces/keeper/msg_server.go:279]: What if is the subspace owner that wants to change the permissions of the group and is also part of it?

**Review by manu0466 — CHANGES_REQUESTED** [review]

File: x/subspaces/keeper/msg_server.go (hunk 1 above)
Comment (RiccardoM) [reply x/subspaces/keeper/msg_server.go]: Thanks for pointing this out @manu0466. I've updated the code to always allow the owner to set any group permission

**Review by bragaz — APPROVED** [review]

**Comment by manu0466** [comment]
@RiccardoM I think that there is another problem.
From spec a user can't give to other users the `PermissionSetPermissions`, this can be baypassed in this 2 cases: 

## Case 1
A user have the `PermissionSetPermissions` the he can just call `SetUserPermissions` and give to another user the `PermissionSetPermissions`

## Case 2
- User **UA** have the `PermissionManageGroups`, then he create a new group with the defaultPermissions = `PermissionSetPermissions`
- Then he can add himelf to the group inheriting the `PermissionSetPermissions`

## Case 3
- User **UA** sets default permissions of group **GA** to something that have  `PermissionSetPermissions`, **UA** is not meber of the group.
- Then let's assume that **UA** have also the permissions `PermissionManageGroups` then he can add other users to the group giving them the `PermissionSetPermission`  since all the members inherit the group permissions.

In all cases I don't see any checks that makes sure that only the subspace owner can give the `SetUserPermissions` in `SetUserPermissions`, `CreateUserGroup`, `EditUserGroup`.

**Comment by RiccardoM** [comment]
@manu0466 Let me answer you case by case

> ## Case 1
> A user have the `PermissionSetPermissions` the he can just call `SetUserPermissions` and give to another user the `PermissionSetPermissions`

This is the indented behavior. A subspace owner, which is the only initial user that can set permissions, should be very careful about giving out `PermissionSetPermissions` because the following might happen: 
1. Alice gives `PermissionSetPermissions` to Bob
2. Bob created a new account, named Carl
3. Bob gives `PermissionSetPermissions` to another user 

Unfortunately there is nothing we can do here. We should respect the choice of the subspace owner in this case and just allow them to do whatever they want. They can always decide to be the only one to give out permissions if they prefer. We just give options: it's up to them to use this feature or not. 
 
> ## Case 2
> * User **UA** have the `PermissionManageGroups`, then he create a new group with the defaultPermissions = `PermissionSetPermissions`
> * Then he can add himelf to the group inheriting the `PermissionSetPermissions`

This is actually a bug. I think it's better to open an issue about this and make sure during the handling of `MsgCreateUserGroup` that if `DefaultPermissions` contains `PermissionSetPermissions`, then the group owner must also have `PermissionSetPermissions` to avoid any trouble. 

> ## Case 3
> * User **UA** sets default permissions of group **GA** to something that have  `PermissionSetPermissions`, **UA** is not meber of the group.
> * Then let's assume that **UA** have also the permissions `PermissionManageGroups` then he can add other users to the group giving them the `PermissionSetPermission`  since all the members inherit the group permissions.

Actually, in order to add a user to a group you are required to have `PermissionSetPermissions`, not `PermissionManageGroups`. The latter is only checked when creating a group or editing an existing group. Also, editing user groups does **not** allow to set a new permission for the group itself. Groups permissions are only handled though `MsgSetUserGroupPermissions` which require `PermissionSetPermissions`.

**Review by manu0466 — APPROVED** [review]

**Pull request merged**